### PR TITLE
adjust persistent buffer size if can be projected to immediate producers

### PR DESCRIPTION
**Pull request by liqiangxl**
Issue: #1386
In normalization scheduler, if `project_persistent_buffers = True`, `PersistentBufferProjector` will project persistent tvs to inputs, otherwise it trys to project persistent tvs to its immediate persistent producers.
However, the buffer size was over estimated in `scheduler_utils::persistentBufferSize`. This overestimation occurs because the potential to project persistent tvs onto their immediate persistent producers isn't factored into the calculations.

Fix:
This PR fixed this overestimation by taking out the size of buffers which can be projected to its producers.

Influence:
The buffer size is corrected. Lead to smaller maxregcount generated by the heuristics, but no influence on generated code of the cpp benchmarks/tests (checked with `!build --diff --diff-bench`) since the current persistent batch size depends on elements count rather than buffer size.

### CODE REVIEW

**Comment by liqiangxl** [comment]
!build --diff --diff-bench

**Comment by naoyam** [comment]
!build --diff --diff-bench

File: csrc/scheduler/utils.cpp [hunk 1]
```diff
@@ -915,6 +932,19 @@ PersistentBufferSizeReturn persistentBufferSize(
         std::max(max_proj_persistence_size, projected_buffer_size);
   }
 
+  // reduce max_persistence_size if a persistent buffer can be projected to
+  // other persistent tvs.
+  std::unordered_set<TensorView*> persistent_buffer_set(
```
Comment (naoyam) [inline csrc/scheduler/utils.cpp:937]: Persistent buffers are scoped, and the buffer size is the largest among multiple scopes. This reduction of the projection to immediate producers needs to be done with each scope.

Try to create an example with multiple scopes.

**Review by naoyam — CHANGES_REQUESTED** [review]

**Comment by liqiangxl** [comment]
!build --diff --diff-bench

File: csrc/scheduler/utils.cpp (hunk 1 above)
Comment (liqiangxl) [reply csrc/scheduler/utils.cpp]: Good point. In the newly added test case, the largest buffer size comes from a scope which doesn't have a projecatable buffer. The other scope has a projecatable buffer. So the previous implementation overestimated the buffer size. In the revised version, the `persistent_mask` is used in the buffer size calculation. It is set to `true` when 
```
    if (!canProjectToPersistentProducer(
            buffer, producers, persistent_buffer_set)) {
      persistent_mask[buffer_i] = true;
    }
```
when the mask is `false`, the size is not counted.

**Comment by liqiangxl** [comment]
For `nvfuser-ci/jit_codegen_diff_bench_17_5/5 — Failing (w/msg) `, the different PTX is expected since the maxregcount passed to NVRTC is changed.

File: test/test_gpu3.cpp [hunk 2]
```diff
@@ -9409,6 +9413,90 @@ TEST_F(NVFuserTest, LoweringHook) {
...
+  auto t0 = at::randn({batch_size, hidden_size}, options);
+  auto t1 = at::randn({batch_size, hidden_size}, options);
+  auto t2 = at::randn({batch_size, hidden_size}, options);
+  std::vector<c10::IValue> inputs{t0, t1, t2};
+
+  // The persistent buffers in this fusion are: tv3, tv7, tv12, and tv17. Note
+  // that tv7 can be projected back to its producer, tv3. When calculating the
+  // total size of persistent buffers ([persistent_buffer_size]), it's important
+  // to consider the active scopes of these buffers. Simply subtracting the
+  // buffer size of tv7 from the total may lead to an overestimation. This is
```
[the remark this answers is not in the document]
Comment (liqiangxl) [reply test/test_gpu3.cpp:9471]: oh, you are right, underestimation.

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**